aws-signer/src/main/scala/org/http4s/aws/middleware/AwsSigner.scala

```diff
@@ -0,0 +1,123 @@
+package org.http4s.aws.middleware
+
+import java.util.Date
+
+import cats.data.Kleisli
+import cats.effect.Sync
+import cats.implicits._
+import fs2.Stream
+import org.http4s.client.Client
+import org.http4s.{Header, Request}
+import java.nio.charset.StandardCharsets
+
+import org.http4s.internal.BytesToHex.bytesToHex
+
+object AwsSigner {
+  def apply[F[_]: Sync](id: String, secret: String, zone: String, service: String)(
+      client: Client[F]): Client[F] = {
+    val signer = new AwsSigner(Key(id, secret), zone, service)
+    client.copy(open = Kleisli { req =>
+      signer[F](req, Sync[F].delay(new Date)).flatMap(client.open(_))
+    })
+  }
+
+  private final case class Key(id: String, secret: String)
+}
+
+/** Implements AWS request signing v4 as an http4s middleware.
+  * Signing described here: https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
+  */
+class AwsSigner(key: AwsSigner.Key, zone: String, service: String) {
+  val Method = "AWS4-HMAC-SHA256"
+  val Charset = StandardCharsets.UTF_8
+
+  private def dateFormat(s: String) = {
+    val f = new java.text.SimpleDateFormat(s)
+    f.setTimeZone(java.util.TimeZone.getTimeZone("UTC"))
+    f
+  }
+
+  val fullDateFormat = dateFormat("YYYYMMdd'T'HHmmss'Z'")
+  val shortDateFormat = dateFormat("YYYYMMdd")
```

Comment:
these are not thread safe

Reply:
Will be changing to `Instant`.

In what way are they not thread safe btw?

Reply:
> Date formats are not synchronized. It is recommended to create separate format instances for each thread. If multiple threads access a format concurrently, it must be synchronized externally.

[Reference](https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html)